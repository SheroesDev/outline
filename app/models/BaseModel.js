// @flow
import { set, observable } from 'mobx';

export default class BaseModel {
  @observable id: string;
  @observable isSaving: boolean;
  store: *;

  constructor(fields: Object, store: *) {
    set(this, fields);
    this.store = store;
  }

  save = async params => {
    this.isSaving = true;

    try {
      // ensure that the id is passed if the document has one
      if (params) params = { ...params, id: this.id };
      await this.store.save(params || this.toJS());

      // if saving is successful set the new values on the model itself
      if (params) set(this, params);
      return this;
    } finally {
      this.isSaving = false;
    }
  };

  fetch = (options: *) => {
    return this.store.fetch(this.id, options);
  };

  refresh = () => {
    return this.fetch({ force: true });
  };

  delete = async () => {
    this.isSaving = true;
    try {
      return await this.store.delete(this);
    } finally {
      this.isSaving = false;
    }
  };

  toJS = () => {
    return { ...this };
  };
}
