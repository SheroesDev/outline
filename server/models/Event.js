// @flow
import { DataTypes, sequelize } from '../sequelize';

const Event = sequelize.define('event', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  name: DataTypes.STRING,
  data: DataTypes.JSONB,
});

Event.associate = models => {
  Event.belongsTo(models.User, {
    as: 'user',
    foreignKey: 'userId',
  });
  Event.belongsTo(models.Collection, {
    as: 'collection',
    foreignKey: 'collectionId',
  });
  Event.belongsTo(models.Collection, {
    as: 'document',
    foreignKey: 'documentId',
  });
  Event.belongsTo(models.Team, {
    as: 'team',
    foreignKey: 'teamId',
  });
};

export default Event;
