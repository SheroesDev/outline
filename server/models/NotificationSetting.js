// @flow
import { DataTypes, sequelize } from '../sequelize';

const NotificationSetting = sequelize.define(
  'notification_setting',
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    event: {
      type: DataTypes.STRING,
    },
  },
  {
    timestamps: true,
    updatedAt: false,
  }
);

NotificationSetting.associate = models => {
  NotificationSetting.belongsTo(models.User, {
    as: 'user',
    foreignKey: 'userId',
  });
  NotificationSetting.belongsTo(models.Team, {
    as: 'team',
    foreignKey: 'teamId',
  });
};

export default NotificationSetting;
