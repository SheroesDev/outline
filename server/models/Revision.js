// @flow
import { DataTypes, sequelize } from '../sequelize';

const Revision = sequelize.define('revision', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true,
  },
  editorVersion: DataTypes.STRING,
  title: DataTypes.STRING,
  text: DataTypes.TEXT,
});

Revision.associate = models => {
  Revision.belongsTo(models.Document, {
    as: 'document',
    foreignKey: 'documentId',
    onDelete: 'cascade',
  });
  Revision.belongsTo(models.User, {
    as: 'user',
    foreignKey: 'userId',
  });
  Revision.addScope(
    'defaultScope',
    {
      include: [{ model: models.User, as: 'user', paranoid: false }],
    },
    { override: true }
  );
};

export default Revision;
