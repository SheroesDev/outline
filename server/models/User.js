// @flow
import crypto from 'crypto';
import uuid from 'uuid';
import JWT from 'jsonwebtoken';
import subMinutes from 'date-fns/sub_minutes';
import { DataTypes, sequelize, encryptedFields } from '../sequelize';
import { publicS3Endpoint, uploadToS3FromUrl } from '../utils/s3';
import { sendEmail } from '../mailer';
import { Star, NotificationSetting, ApiKey } from '.';

const User = sequelize.define(
  'user',
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    email: { type: DataTypes.STRING },
    username: { type: DataTypes.STRING },
    name: DataTypes.STRING,
    avatarUrl: { type: DataTypes.STRING, allowNull: true },
    isAdmin: DataTypes.BOOLEAN,
    service: { type: DataTypes.STRING, allowNull: true },
    serviceId: { type: DataTypes.STRING, allowNull: true, unique: true },
    slackData: DataTypes.JSONB,
    jwtSecret: encryptedFields.vault('jwtSecret'),
    lastActiveAt: DataTypes.DATE,
    lastActiveIp: { type: DataTypes.STRING, allowNull: true },
    lastSignedInAt: DataTypes.DATE,
    lastSignedInIp: { type: DataTypes.STRING, allowNull: true },
    suspendedAt: DataTypes.DATE,
    suspendedById: DataTypes.UUID,
  },
  {
    paranoid: true,
    getterMethods: {
      isSuspended() {
        return !!this.suspendedAt;
      },
    },
    indexes: [
      {
        fields: ['email'],
      },
    ],
  }
);

// Class methods
User.associate = models => {
  User.hasMany(models.ApiKey, { as: 'apiKeys' });
  User.hasMany(models.Document, { as: 'documents' });
  User.hasMany(models.View, { as: 'views' });
};

// Instance methods
User.prototype.updateActiveAt = function(ip) {
  const fiveMinutesAgo = subMinutes(new Date(), 5);

  // ensure this is updated only every few minutes otherwise
  // we'll be constantly writing to the DB as API requests happen
  if (this.lastActiveAt < fiveMinutesAgo) {
    this.lastActiveAt = new Date();
    this.lastActiveIp = ip;
    return this.save({ hooks: false });
  }
};

User.prototype.updateSignedIn = function(ip) {
  this.lastSignedInAt = new Date();
  this.lastSignedInIp = ip;
  return this.save({ hooks: false });
};

User.prototype.getJwtToken = function() {
  return JWT.sign({ id: this.id }, this.jwtSecret);
};

const uploadAvatar = async model => {
  const endpoint = publicS3Endpoint();

  if (model.avatarUrl && !model.avatarUrl.startsWith(endpoint)) {
    const newUrl = await uploadToS3FromUrl(
      model.avatarUrl,
      `avatars/${model.id}/${uuid.v4()}`
    );
    if (newUrl) model.avatarUrl = newUrl;
  }
};

const setRandomJwtSecret = model => {
  model.jwtSecret = crypto.randomBytes(64).toString('hex');
};

const removeIdentifyingInfo = async model => {
  await ApiKey.destroy({ where: { userId: model.id } });
  await Star.destroy({ where: { userId: model.id } });

  model.email = '';
  model.name = 'Unknown';
  model.avatarUrl = '';
  model.serviceId = null;
  model.username = null;
  model.slackData = null;
  model.lastActiveIp = null;
  model.lastSignedInIp = null;

  // this shouldn't be needed once this issue is resolved:
  // https://github.com/sequelize/sequelize/issues/9318
  await model.save({ hooks: false });
};

const checkLastAdmin = async model => {
  const teamId = model.teamId;

  if (model.isAdmin) {
    const userCount = await User.count({ where: { teamId } });
    const adminCount = await User.count({ where: { isAdmin: true, teamId } });

    if (userCount > 1 && adminCount <= 1) {
      throw new Error(
        'Cannot delete account as only admin. Please transfer admin permissions to another user and try again.'
      );
    }
  }
};

User.beforeDestroy(checkLastAdmin);
User.beforeDestroy(removeIdentifyingInfo);
User.beforeSave(uploadAvatar);
User.beforeCreate(setRandomJwtSecret);
User.afterCreate(user => sendEmail('welcome', user.email));

// By default when a user signs up we subscribe them to email notifications
// when documents they created are edited by other team members.
User.afterCreate((user, options) =>
  NotificationSetting.findOrCreate({
    where: {
      userId: user.id,
      teamId: user.teamId,
      event: 'documents.update',
    },
    transaction: options.transaction,
  })
);

export default User;
