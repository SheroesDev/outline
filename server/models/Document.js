// @flow
import slug from 'slug';
import _ from 'lodash';
import randomstring from 'randomstring';
import MarkdownSerializer from 'slate-md-serializer';
import Plain from 'slate-plain-serializer';
import Sequelize from 'sequelize';

import isUUID from 'validator/lib/isUUID';
import { Collection } from '../models';
import { DataTypes, sequelize } from '../sequelize';
import events from '../events';
import parseTitle from '../../shared/utils/parseTitle';
import Revision from './Revision';

const Op = Sequelize.Op;
const Markdown = new MarkdownSerializer();
const URL_REGEX = /^[a-zA-Z0-9-]*-([a-zA-Z0-9]{10,15})$/;
const DEFAULT_TITLE = 'Untitled document';

// $FlowIssue invalid flow-typed
slug.defaults.mode = 'rfc3986';
const slugify = text =>
  slug(text, {
    remove: /[.]/g,
  });

const createRevision = (doc, options = {}) => {
  if (options.autosave) return;

  return Revision.create({
    title: doc.title,
    text: doc.text,
    userId: doc.lastModifiedById,
    documentId: doc.id,
  });
};

const createUrlId = doc => {
  return (doc.urlId = doc.urlId || randomstring.generate(10));
};

const beforeSave = async doc => {
  const { emoji, title } = parseTitle(doc.text);

  // emoji in the title is split out for easier display
  doc.emoji = emoji;

  // ensure document has a title
  if (!title) {
    doc.title = DEFAULT_TITLE;
    doc.text = doc.text.replace(/^.*$/m, `# ${DEFAULT_TITLE}`);
  }

  // calculate collaborators
  let ids = [];
  if (doc.id) {
    ids = await Revision.findAll({
      attributes: [[DataTypes.literal('DISTINCT "userId"'), 'userId']],
      where: {
        documentId: doc.id,
      },
    }).map(rev => rev.userId);
  }

  // add the current user as revision hasn't been generated yet
  ids.push(doc.lastModifiedById);
  doc.collaboratorIds = _.uniq(ids);

  // increment revision
  doc.revisionCount += 1;

  return doc;
};

const Document = sequelize.define(
  'document',
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    urlId: { type: DataTypes.STRING, primaryKey: true },
    private: { type: DataTypes.BOOLEAN, defaultValue: true },
    title: DataTypes.STRING,
    text: DataTypes.TEXT,
    revisionCount: { type: DataTypes.INTEGER, defaultValue: 0 },
    publishedAt: DataTypes.DATE,
    parentDocumentId: DataTypes.UUID,
    collaboratorIds: DataTypes.ARRAY(DataTypes.UUID),
  },
  {
    paranoid: true,
    hooks: {
      beforeValidate: createUrlId,
      beforeCreate: beforeSave,
      beforeUpdate: beforeSave,
      afterCreate: createRevision,
      afterUpdate: createRevision,
    },
  }
);

// Class methods

Document.associate = models => {
  Document.belongsTo(models.Collection, {
    as: 'collection',
    foreignKey: 'atlasId',
    onDelete: 'cascade',
  });
  Document.belongsTo(models.Team, {
    as: 'team',
    foreignKey: 'teamId',
  });
  Document.belongsTo(models.User, {
    as: 'createdBy',
    foreignKey: 'createdById',
  });
  Document.belongsTo(models.User, {
    as: 'updatedBy',
    foreignKey: 'lastModifiedById',
  });
  Document.belongsTo(models.User, {
    as: 'pinnedBy',
    foreignKey: 'pinnedById',
  });
  Document.hasMany(models.Revision, {
    as: 'revisions',
    onDelete: 'cascade',
  });
  Document.hasMany(models.Star, {
    as: 'starred',
  });
  Document.hasMany(models.View, {
    as: 'views',
  });
  Document.addScope(
    'defaultScope',
    {
      include: [
        { model: models.Collection, as: 'collection' },
        { model: models.User, as: 'createdBy', paranoid: false },
        { model: models.User, as: 'updatedBy', paranoid: false },
      ],
      where: {
        publishedAt: {
          // $FlowFixMe
          [Op.ne]: null,
        },
      },
    },
    { override: true }
  );
  Document.addScope('withUnpublished', {
    include: [
      { model: models.Collection, as: 'collection' },
      { model: models.User, as: 'createdBy', paranoid: false },
      { model: models.User, as: 'updatedBy', paranoid: false },
    ],
  });
  Document.addScope('withViews', userId => ({
    include: [
      { model: models.View, as: 'views', where: { userId }, required: false },
    ],
  }));
  Document.addScope('withStarred', userId => ({
    include: [
      { model: models.Star, as: 'starred', where: { userId }, required: false },
    ],
  }));
};

Document.findById = async id => {
  const scope = Document.scope('withUnpublished');

  if (isUUID(id)) {
    return scope.findOne({
      where: { id },
    });
  } else if (id.match(URL_REGEX)) {
    return scope.findOne({
      where: {
        urlId: id.match(URL_REGEX)[1],
      },
    });
  }
};

Document.searchForUser = async (
  user,
  query,
  options = {}
): Promise<Document[]> => {
  const limit = options.limit || 15;
  const offset = options.offset || 0;

  const sql = `
        SELECT *, ts_rank(documents."searchVector", plainto_tsquery('english', :query)) as "searchRanking" FROM documents
        WHERE "searchVector" @@ plainto_tsquery('english', :query) AND
          "teamId" = '${user.teamId}'::uuid AND
          "deletedAt" IS NULL
        ORDER BY "searchRanking" DESC
        LIMIT :limit OFFSET :offset;
        `;

  const results = await sequelize.query(sql, {
    replacements: {
      query,
      limit,
      offset,
    },
    model: Document,
  });
  const ids = results.map(document => document.id);

  // Second query to get views for the data
  const withViewsScope = { method: ['withViews', user.id] };
  const documents = await Document.scope(
    'defaultScope',
    withViewsScope
  ).findAll({
    where: { id: ids },
  });

  // Order the documents in the same order as the first query
  return _.sortBy(documents, doc => ids.indexOf(doc.id));
};

// Hooks

Document.addHook('beforeSave', async model => {
  if (!model.publishedAt) return;

  const collection = await Collection.findById(model.atlasId);
  if (collection.type !== 'atlas') return;

  await collection.updateDocument(model);
  model.collection = collection;
});

Document.addHook('afterCreate', async model => {
  if (!model.publishedAt) return;

  const collection = await Collection.findById(model.atlasId);
  if (collection.type !== 'atlas') return;

  await collection.addDocumentToStructure(model);
  model.collection = collection;

  events.add({ name: 'documents.create', model });
  return model;
});

Document.addHook('afterDestroy', model =>
  events.add({ name: 'documents.delete', model })
);

// Instance methods

Document.prototype.publish = async function() {
  if (this.publishedAt) return this.save();

  const collection = await Collection.findById(this.atlasId);
  if (collection.type !== 'atlas') return this.save();

  await collection.addDocumentToStructure(this);

  this.publishedAt = new Date();
  await this.save();
  this.collection = collection;

  events.add({ name: 'documents.publish', model: this });
  return this;
};

Document.prototype.getTimestamp = function() {
  return Math.round(new Date(this.updatedAt).getTime() / 1000);
};

Document.prototype.getSummary = function() {
  const value = Markdown.deserialize(this.text);
  const plain = Plain.serialize(value);
  const lines = _.compact(plain.split('\n'));
  return lines.length >= 1 ? lines[1] : '';
};

Document.prototype.getUrl = function() {
  const slugifiedTitle = slugify(this.title);
  return `/doc/${slugifiedTitle}-${this.urlId}`;
};

Document.prototype.toJSON = function() {
  // Warning: only use for new documents as order of children is
  // handled in the collection's documentStructure
  return {
    id: this.id,
    title: this.title,
    url: this.getUrl(),
    children: [],
  };
};

export default Document;
