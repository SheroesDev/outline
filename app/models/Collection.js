// @flow
import { pick } from 'lodash';
import { action, computed } from 'mobx';
import BaseModel from 'models/BaseModel';
import Document from 'models/Document';
import { client } from 'utils/ApiClient';
import type { NavigationNode } from 'types';

export default class Collection extends BaseModel {
  isSaving: boolean;

  id: string;
  name: string;
  description: string;
  color: string;
  type: 'atlas' | 'journal';
  documents: NavigationNode[];
  createdAt: ?string;
  updatedAt: ?string;
  url: string;

  @computed
  get isEmpty(): boolean {
    return this.documents.length === 0;
  }

  @computed
  get documentIds(): string[] {
    const results = [];
    const travelDocuments = (documentList, path) =>
      documentList.forEach(document => {
        results.push(document.id);
        travelDocuments(document.children);
      });

    travelDocuments(this.documents);
    return results;
  }

  @action
  updateDocument(document: Document) {
    const travelDocuments = (documentList, path) =>
      documentList.forEach(d => {
        if (d.id === document.id) {
          d.title = document.title;
          d.url = document.url;
        } else {
          travelDocuments(d.children);
        }
      });

    travelDocuments(this.documents);
  }

  toJS = () => {
    return pick(this, ['name', 'color', 'description']);
  };

  export = () => {
    return client.post('/collections.export', { id: this.id });
  };
}
