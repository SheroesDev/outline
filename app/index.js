// @flow
import * as React from 'react';
import { render } from 'react-dom';
import { Provider } from 'mobx-react';
import { ThemeProvider } from 'styled-components';
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Redirect,
} from 'react-router-dom';

import stores from 'stores';
import theme from 'shared/styles/theme';
import globalStyles from 'shared/styles/globals';
import 'shared/styles/prism.css';

import Home from 'scenes/Home';
import Dashboard from 'scenes/Dashboard';
import Starred from 'scenes/Starred';
import Drafts from 'scenes/Drafts';
import Collection from 'scenes/Collection';
import Document from 'scenes/Document';
import Search from 'scenes/Search';
import Settings from 'scenes/Settings';
import Details from 'scenes/Settings/Details';
import People from 'scenes/Settings/People';
import Slack from 'scenes/Settings/Slack';
import Shares from 'scenes/Settings/Shares';
import Tokens from 'scenes/Settings/Tokens';
import Export from 'scenes/Settings/Export';
import Error404 from 'scenes/Error404';

import ErrorBoundary from 'components/ErrorBoundary';
import ScrollToTop from 'components/ScrollToTop';
import ScrollToAnchor from 'components/ScrollToAnchor';
import Layout from 'components/Layout';
import Auth from 'components/Auth';
import RouteSidebarHidden from 'components/RouteSidebarHidden';

import { matchDocumentSlug } from 'utils/routeHelpers';

let DevTools;
if (__DEV__) {
  DevTools = require('mobx-react-devtools').default; // eslint-disable-line global-require
}

const notFoundSearch = () => <Search notFound />;
const DocumentNew = () => <Document newDocument />;
const RedirectDocument = ({ match }: { match: Object }) => (
  <Redirect to={`/doc/${match.params.documentSlug}`} />
);

globalStyles();

const element = document.getElementById('root');

if (element) {
  render(
    <React.Fragment>
      <ThemeProvider theme={theme}>
        <ErrorBoundary>
          <Provider {...stores}>
            <Router>
              <ScrollToTop>
                <ScrollToAnchor>
                  <Switch>
                    <Route exact path="/" component={Home} />
                    <Route exact path="/share/:shareId" component={Document} />
                    <Auth>
                      <Layout>
                        <Switch>
                          <Route
                            exact
                            path="/dashboard"
                            component={Dashboard}
                          />
                          <Route exact path="/starred" component={Starred} />
                          <Route exact path="/drafts" component={Drafts} />
                          <Route exact path="/settings" component={Settings} />
                          <Route
                            exact
                            path="/settings/details"
                            component={Details}
                          />
                          <Route
                            exact
                            path="/settings/people"
                            component={People}
                          />
                          <Route
                            exact
                            path="/settings/shares"
                            component={Shares}
                          />
                          <Route
                            exact
                            path="/settings/tokens"
                            component={Tokens}
                          />
                          <Route
                            exact
                            path="/settings/integrations/slack"
                            component={Slack}
                          />
                          <Route
                            exact
                            path="/settings/export"
                            component={Export}
                          />
                          <Route
                            exact
                            path="/collections/:id"
                            component={Collection}
                          />
                          <Route
                            exact
                            path={`/d/${matchDocumentSlug}`}
                            component={RedirectDocument}
                          />
                          <Route
                            exact
                            path={`/doc/${matchDocumentSlug}`}
                            component={Document}
                          />
                          <Route
                            exact
                            path={`/doc/${matchDocumentSlug}/move`}
                            component={Document}
                          />
                          <Route exact path="/search" component={Search} />
                          <Route
                            exact
                            path="/search/:query"
                            component={Search}
                          />
                          <Route path="/404" component={Error404} />
                          <RouteSidebarHidden
                            exact
                            path={`/doc/${matchDocumentSlug}/edit`}
                            component={Document}
                          />
                          <RouteSidebarHidden
                            exact
                            path="/collections/:id/new"
                            component={DocumentNew}
                          />
                          <Route component={notFoundSearch} />
                        </Switch>
                      </Layout>
                    </Auth>
                  </Switch>
                </ScrollToAnchor>
              </ScrollToTop>
            </Router>
          </Provider>
        </ErrorBoundary>
      </ThemeProvider>
      {DevTools && <DevTools position={{ bottom: 0, right: 0 }} />}
    </React.Fragment>,
    element
  );
}

window.addEventListener('load', async () => {
  // installation does not use Google Analytics, or tracking is blocked on client
  // no point loading the rest of the analytics bundles
  if (!process.env.GOOGLE_ANALYTICS_ID || !window.ga) return;

  // https://github.com/googleanalytics/autotrack/issues/137#issuecomment-305890099
  await import('autotrack/autotrack.js');

  window.ga('require', 'outboundLinkTracker');
  window.ga('require', 'urlChangeTracker');
});
