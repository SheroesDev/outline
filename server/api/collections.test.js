/* eslint-disable flowtype/require-valid-file-annotation */
import TestServer from 'fetch-test-server';
import app from '..';
import { flushdb, seed } from '../test/support';
import { buildUser } from '../test/factories';
import { Collection } from '../models';
const server = new TestServer(app.callback());

beforeEach(flushdb);
afterAll(server.close);

describe('#collections.list', async () => {
  it('should require authentication', async () => {
    const res = await server.post('/api/collections.list');
    const body = await res.json();

    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it('should return collections', async () => {
    const { user, collection } = await seed();
    const res = await server.post('/api/collections.list', {
      body: { token: user.getJwtToken() },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.length).toEqual(1);
    expect(body.data[0].id).toEqual(collection.id);
  });
});

describe('#collections.export', async () => {
  it('should require authentication', async () => {
    const res = await server.post('/api/collections.export');
    const body = await res.json();

    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it('should return success', async () => {
    const { user, collection } = await seed();
    const res = await server.post('/api/collections.export', {
      body: { token: user.getJwtToken(), id: collection.id },
    });

    expect(res.status).toEqual(200);
  });
});

describe('#collections.exportAll', async () => {
  it('should require authentication', async () => {
    const res = await server.post('/api/collections.exportAll');
    const body = await res.json();

    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it('should require authorization', async () => {
    const user = await buildUser();
    const res = await server.post('/api/collections.exportAll', {
      body: { token: user.getJwtToken() },
    });
    expect(res.status).toEqual(403);
  });

  it('should return success', async () => {
    const { admin } = await seed();
    const res = await server.post('/api/collections.exportAll', {
      body: { token: admin.getJwtToken() },
    });

    expect(res.status).toEqual(200);
  });
});

describe('#collections.info', async () => {
  it('should return collection', async () => {
    const { user, collection } = await seed();
    const res = await server.post('/api/collections.info', {
      body: { token: user.getJwtToken(), id: collection.id },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.id).toEqual(collection.id);
  });

  it('should require authentication', async () => {
    const res = await server.post('/api/collections.info');
    const body = await res.json();

    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it('should require authorization', async () => {
    const { collection } = await seed();
    const user = await buildUser();
    const res = await server.post('/api/collections.info', {
      body: { token: user.getJwtToken(), id: collection.id },
    });
    expect(res.status).toEqual(403);
  });
});

describe('#collections.create', async () => {
  it('should require authentication', async () => {
    const res = await server.post('/api/collections.create');
    const body = await res.json();

    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it('should create collection', async () => {
    const { user } = await seed();
    const res = await server.post('/api/collections.create', {
      body: { token: user.getJwtToken(), name: 'Test', type: 'atlas' },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.data.id).toBeTruthy();
    expect(body.data.name).toBe('Test');
  });
});

describe('#collections.delete', async () => {
  it('should require authentication', async () => {
    const res = await server.post('/api/collections.delete');
    const body = await res.json();

    expect(res.status).toEqual(401);
    expect(body).toMatchSnapshot();
  });

  it('should require authorization', async () => {
    const { collection } = await seed();
    const user = await buildUser();
    const res = await server.post('/api/collections.delete', {
      body: { token: user.getJwtToken(), id: collection.id },
    });
    expect(res.status).toEqual(403);
  });

  it('should not delete last collection', async () => {
    const { user, collection } = await seed();
    const res = await server.post('/api/collections.delete', {
      body: { token: user.getJwtToken(), id: collection.id },
    });
    expect(res.status).toEqual(400);
  });

  it('should delete collection', async () => {
    const { user, collection } = await seed();
    await Collection.create({
      name: 'Blah',
      urlId: 'blah',
      teamId: user.teamId,
      creatorId: user.id,
      type: 'atlas',
    });

    const res = await server.post('/api/collections.delete', {
      body: { token: user.getJwtToken(), id: collection.id },
    });
    const body = await res.json();

    expect(res.status).toEqual(200);
    expect(body.success).toBe(true);
  });
});
