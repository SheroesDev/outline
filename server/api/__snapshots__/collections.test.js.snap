// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#collections.create should require authentication 1`] = `
Object {
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#collections.delete should require authentication 1`] = `
Object {
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#collections.export should require authentication 1`] = `
Object {
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#collections.exportAll should require authentication 1`] = `
Object {
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#collections.info should require authentication 1`] = `
Object {
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;

exports[`#collections.list should require authentication 1`] = `
Object {
  "error": "authentication_required",
  "message": "Authentication required",
  "ok": false,
  "status": 401,
}
`;
